File: TranslationFiestaGo/internal/data/services/local_service_client.go
package services

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"time"
	"translationfiestago/internal/domain/entities"
	"translationfiestago/internal/utils"
)

const defaultLocalServiceURL = "http://127.0.0.1:5055"
const defaultLocalServiceScript = "TranslationFiestaLocal/local_service.py"

type LocalServiceClient struct {
	httpClient *utils.HTTPClient
	logger     *utils.Logger
	baseURL    string
	autoStart  bool
	startMu    sync.Mutex
	started    bool
}

type localServiceError struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

type localTranslateResponse struct {
	TranslatedText string             `json:"translated_text"`
	SourceLang     string             `json:"source_lang"`
	TargetLang     string             `json:"target_lang"`
	Backend        string             `json:"backend"`
	Error          *localServiceError `json:"error,omitempty"`
}

type localHealthResponse struct {
	Status  string `json:"status"`
	Backend string `json:"backend"`
}

func NewLocalServiceClient(httpClient *utils.HTTPClient) *LocalServiceClient {
	client := &LocalServiceClient{
		httpClient: httpClient,
		logger:     utils.GetLogger(),
	}
	client.refreshFromEnv()
	return client
}

func (c *LocalServiceClient) Translate(ctx context.Context, text, sourceLang, targetLang string) (*entities.TranslationResult, error) {
	c.refreshFromEnv()
	if err := c.ensureAvailable(ctx); err != nil {
		return nil, err
	}

	payload := map[string]string{
		"text":        text,
		"source_lang": sourceLang,
		"target_lang": targetLang,
	}

	resp, err := c.httpClient.Post(ctx, c.baseURL+"/translate", payload)
	if err != nil {
		return nil, fmt.Errorf("local service request failed: %w", err)
	}

	if resp.StatusCode() != 200 {
		return nil, c.parseError(resp.Body(), resp.StatusCode())
	}

	var parsed localTranslateResponse
	if err := json.Unmarshal(resp.Body(), &parsed); err != nil {
		return nil, fmt.Errorf("local service response invalid: %w", err)
	}
	if parsed.Error != nil {
		return nil, fmt.Errorf("local service error: %s", parsed.Error.Message)
	}

	translated := strings.TrimSpace(parsed.TranslatedText)
	if translated == "" {
		return nil, fmt.Errorf("local service returned empty translation")
	}

	resSource := parsed.SourceLang
	if resSource == "" {
		resSource = sourceLang
	}
	resTarget := parsed.TargetLang
	if resTarget == "" {
		resTarget = targetLang
	}

	return &entities.TranslationResult{
		TranslatedText: translated,
		SourceLang:     resSource,
		TargetLang:     resTarget,
		Timestamp:      time.Now(),
	}, nil
}

func (c *LocalServiceClient) ModelsStatus(ctx context.Context) (string, error) {
	c.refreshFromEnv()
	if err := c.ensureAvailable(ctx); err != nil {
		return "", err
	}
	resp, err := c.httpClient.Get(ctx, c.baseURL+"/models")
	if err != nil {
		return "", fmt.Errorf("local models status failed: %w", err)
	}
	if resp.StatusCode() != 200 {
		return "", c.parseError(resp.Body(), resp.StatusCode())
	}
	return string(resp.Body()), nil
}

func (c *LocalServiceClient) ModelsVerify(ctx context.Context) (string, error) {
	c.refreshFromEnv()
	if err := c.ensureAvailable(ctx); err != nil {
		return "", err
	}
	resp, err := c.httpClient.Post(ctx, c.baseURL+"/models/verify", map[string]string{})
	if err != nil {
		return "", fmt.Errorf("local models verify failed: %w", err)
	}
	if resp.StatusCode() != 200 {
		return "", c.parseError(resp.Body(), resp.StatusCode())
	}
	return string(resp.Body()), nil
}

func (c *LocalServiceClient) ModelsRemove(ctx context.Context) (string, error) {
	c.refreshFromEnv()
	if err := c.ensureAvailable(ctx); err != nil {
		return "", err
	}
	resp, err := c.httpClient.Post(ctx, c.baseURL+"/models/remove", map[string]string{})
	if err != nil {
		return "", fmt.Errorf("local models remove failed: %w", err)
	}
	if resp.StatusCode() != 200 {
		return "", c.parseError(resp.Body(), resp.StatusCode())
	}
	return string(resp.Body()), nil
}

func (c *LocalServiceClient) ModelsInstallDefault(ctx context.Context) (string, error) {
	c.refreshFromEnv()
	if err := c.ensureAvailable(ctx); err != nil {
		return "", err
	}
	resp, err := c.httpClient.Post(ctx, c.baseURL+"/models/install", map[string]string{})
	if err != nil {
		return "", fmt.Errorf("local models install failed: %w", err)
	}
	if resp.StatusCode() != 200 {
		return "", c.parseError(resp.Body(), resp.StatusCode())
	}
	return string(resp.Body()), nil
}

func (c *LocalServiceClient) ensureAvailable(ctx context.Context) error {
	c.refreshFromEnv()
	if err := c.checkHealth(ctx); err == nil {
		return nil
	}
	if !c.autoStart {
		return fmt.Errorf("local service unavailable and autostart disabled")
	}

	if err := c.startLocalService(); err != nil {
		return err
	}

	return c.waitForHealth(ctx, 10, 250*time.Millisecond)
}

func (c *LocalServiceClient) waitForHealth(ctx context.Context, attempts int, delay time.Duration) error {
	var lastErr error
	for i := 0; i < attempts; i++ {
		if err := c.checkHealth(ctx); err == nil {
			return nil
		} else {
			lastErr = err
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(delay):
		}
	}
	if lastErr != nil {
		return lastErr
	}
	return fmt.Errorf("local service did not become healthy")
}

func (c *LocalServiceClient) checkHealth(ctx context.Context) error {
	c.refreshFromEnv()
	resp, err := c.httpClient.Get(ctx, c.baseURL+"/health")
	if err != nil {
		return fmt.Errorf("local service health check failed: %w", err)
	}
	if resp.StatusCode() != 200 {
		return fmt.Errorf("local service health check HTTP %d", resp.StatusCode())
	}
	var payload localHealthResponse
	if err := json.Unmarshal(resp.Body(), &payload); err != nil {
		return fmt.Errorf("local service health response invalid: %w", err)
	}
	if strings.ToLower(payload.Status) != "ok" {
		return fmt.Errorf("local service status %s", payload.Status)
	}
	return nil
}

func (c *LocalServiceClient) refreshFromEnv() {
	baseURL := strings.TrimSpace(os.Getenv("TF_LOCAL_URL"))
	if baseURL == "" {
		baseURL = defaultLocalServiceURL
	}
	c.baseURL = strings.TrimRight(baseURL, "/")

	autoStart := true
	if raw, ok := os.LookupEnv("TF_LOCAL_AUTOSTART"); ok {
		value := strings.ToLower(strings.TrimSpace(raw))
		if value == "0" || value == "false" || value == "no" {
			autoStart = false
		}
	}
	c.autoStart = autoStart
}

func (c *LocalServiceClient) startLocalService() error {
	c.startMu.Lock()
	defer c.startMu.Unlock()

	if c.started {
		return nil
	}

	scriptPath := strings.TrimSpace(os.Getenv("TF_LOCAL_SCRIPT"))
	if scriptPath == "" {
		scriptPath = defaultLocalServiceScript
	}
	scriptPath = filepath.Clean(scriptPath)
	if !filepath.IsAbs(scriptPath) {
		if cwd, err := os.Getwd(); err == nil {
			scriptPath = filepath.Join(cwd, scriptPath)
		}
	}

	python := strings.TrimSpace(os.Getenv("PYTHON"))
	if python == "" {
		python = "python"
	}

	cmd := exec.Command(python, scriptPath, "serve")
	cmd.Env = os.Environ()
	cmd.Dir = filepath.Dir(scriptPath)
	if err := cmd.Start(); err != nil {
		return fmt.Errorf("failed to start local service: %w", err)
	}
	_ = cmd.Process.Release()
	c.started = true
	c.logger.Info("Local service start requested: %s", scriptPath)
	return nil
}

func (c *LocalServiceClient) parseError(body []byte, statusCode int) error {
	var payload struct {
		Error *localServiceError `json:"error"`
	}
	if err := json.Unmarshal(body, &payload); err == nil && payload.Error != nil {
		return fmt.Errorf("local service error (%s): %s", payload.Error.Code, payload.Error.Message)
	}
	return fmt.Errorf("local service HTTP %d", statusCode)
}

File: TranslationFiestaSwift/Sources/TranslationFiestaSwift/Data/Services/LocalServiceClient.swift
import Foundation
import Logging

public struct LocalServiceConfiguration {
    public let baseURL: URL
    public let modelDir: String
    public let autoStart: Bool

    public init(baseURL: URL, modelDir: String, autoStart: Bool) {
        self.baseURL = baseURL
        self.modelDir = modelDir
        self.autoStart = autoStart
    }

    public static func fromEnvironment() -> LocalServiceConfiguration {
        let rawUrl = ProcessInfo.processInfo.environment["TF_LOCAL_URL"]?.trimmingCharacters(in: .whitespacesAndNewlines)
        let urlString = (rawUrl?.isEmpty == false ? rawUrl! : "http://127.0.0.1:5055")
        let url = URL(string: urlString.trimmingCharacters(in: CharacterSet(charactersIn: "/"))) ?? URL(string: "http://127.0.0.1:5055")!

        let rawAuto = ProcessInfo.processInfo.environment["TF_LOCAL_AUTOSTART"]?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let autoStart: Bool
        if let rawAuto, rawAuto == "0" || rawAuto == "false" || rawAuto == "no" {
            autoStart = false
        } else {
            autoStart = true
        }

        let modelDir = ProcessInfo.processInfo.environment["TF_LOCAL_MODEL_DIR"]?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return LocalServiceConfiguration(baseURL: url, modelDir: modelDir, autoStart: autoStart)
    }

    public static func fromSettings(_ settings: LocalModelSettings) -> LocalServiceConfiguration {
        let trimmedUrl = settings.serviceUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        let url = URL(string: trimmedUrl.isEmpty ? "http://127.0.0.1:5055" : trimmedUrl) ?? URL(string: "http://127.0.0.1:5055")!
        return LocalServiceConfiguration(baseURL: url, modelDir: settings.modelDir, autoStart: settings.autoStart)
    }
}

public final class LocalServiceClient {
    private let logger = Logger(label: "LocalServiceClient")
    private let session: URLSession
    private let configuration: LocalServiceConfiguration
    private var started = false
    private let lock = NSLock()

    public init(session: URLSession = .shared, configuration: LocalServiceConfiguration = .fromEnvironment()) {
        self.session = session
        self.configuration = configuration
    }

    public func translate(text: String, source: String, target: String) async throws -> String {
        try await ensureAvailable()

        let payload: [String: String] = [
            "text": text,
            "source_lang": source,
            "target_lang": target
        ]

        let data = try JSONSerialization.data(withJSONObject: payload)
        var request = URLRequest(url: configuration.baseURL.appendingPathComponent("translate"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = data

        let (responseData, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw TranslationError.networkError("Invalid response from local service")
        }
        guard 200...299 ~= http.statusCode else {
            throw TranslationError.networkError("Local service HTTP \(http.statusCode)")
        }

        let json = try JSONSerialization.jsonObject(with: responseData) as? [String: Any]
        if let error = json?["error"] as? [String: Any],
           let message = error["message"] as? String {
            throw TranslationError.networkError(message)
        }
        if let translated = json?["translated_text"] as? String, !translated.isEmpty {
            return translated
        }
        throw TranslationError.networkError("Local service returned empty translation")
    }

    private func ensureAvailable() async throws {
        if try await checkHealth() {
            return
        }
        guard configuration.autoStart else {
            throw TranslationError.networkError("Local service unavailable and autostart disabled")
        }
        startLocalService()
        for _ in 0..<10 {
            if try await checkHealth() {
                return
            }
            try await Task.sleep(nanoseconds: 250_000_000)
        }
        throw TranslationError.networkError("Local service did not become healthy")
    }

    private func checkHealth() async throws -> Bool {
        let url = configuration.baseURL.appendingPathComponent("health")
        let (responseData, response) = try await session.data(from: url)
        guard let http = response as? HTTPURLResponse, 200...299 ~= http.statusCode else {
            return false
        }
        let json = try JSONSerialization.jsonObject(with: responseData) as? [String: Any]
        let status = json?["status"] as? String
        return status?.lowercased() == "ok"
    }

    private func startLocalService() {
        lock.lock()
        defer { lock.unlock() }
        if started { return }

        let scriptEnv = ProcessInfo.processInfo.environment["TF_LOCAL_SCRIPT"]?.trimmingCharacters(in: .whitespacesAndNewlines)
        let scriptPath = (scriptEnv?.isEmpty == false ? scriptEnv! : "TranslationFiestaLocal/local_service.py")
        let pythonEnv = ProcessInfo.processInfo.environment["PYTHON"]?.trimmingCharacters(in: .whitespacesAndNewlines)
        let pythonExe = (pythonEnv?.isEmpty == false ? pythonEnv! : "python")

        let process = Process()
        process.launchPath = pythonExe
        process.arguments = [scriptPath, "serve"]
        var env = ProcessInfo.processInfo.environment
        if !configuration.modelDir.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            env["TF_LOCAL_MODEL_DIR"] = configuration.modelDir.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if let host = configuration.baseURL.host {
            env["TF_LOCAL_HOST"] = host
        }
        if configuration.baseURL.port != nil {
            env["TF_LOCAL_PORT"] = String(configuration.baseURL.port ?? 5055)
        }
        env["TF_LOCAL_AUTOSTART"] = configuration.autoStart ? "1" : "0"
        process.environment = env
        process.standardOutput = nil
        process.standardError = nil
        process.standardInput = nil
        do {
            try process.run()
            started = true
            logger.info("Local service start requested", metadata: ["script": "\(scriptPath)"])
        } catch {
            logger.error("Failed to start local service", metadata: ["error": "\(error)"])
        }
    }

    public func modelsStatus() async throws -> String {
        try await ensureAvailable()
        let url = configuration.baseURL.appendingPathComponent("models")
        let (data, response) = try await session.data(from: url)
        guard let http = response as? HTTPURLResponse, 200...299 ~= http.statusCode else {
            throw TranslationError.networkError("Local models status failed")
        }
        return String(data: data, encoding: .utf8) ?? ""
    }

    public func verifyModels() async throws -> String {
        try await postModelsAction(path: "models/verify")
    }

    public func removeModels() async throws -> String {
        try await postModelsAction(path: "models/remove")
    }

    public func installDefaultModels() async throws -> String {
        try await postModelsAction(path: "models/install")
    }

    private func postModelsAction(path: String) async throws -> String {
        try await ensureAvailable()
        let url = configuration.baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data("{}".utf8)
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, 200...299 ~= http.statusCode else {
            throw TranslationError.networkError("Local models request failed")
        }
        return String(data: data, encoding: .utf8) ?? ""
    }
}
